File: hdd/longest-substring-without-repeating-characters/answer.kt
class Solution {
      fun lengthOfLongestSubstring(s: String): Int {
        var cachedRepeat = 0
        var usingCharacters = mutableSetOf<Char>()
        var cachedChar = ""

        for(char in s){
            cachedChar += char
            if(!usingCharacters.add(char)){
                cachedChar = cachedChar.substringAfter(char)
                usingCharacters = cachedChar.toSet().toMutableSet()
            }

            if(cachedRepeat < cachedChar.length){
                cachedRepeat = cachedChar.length
            }
        }

        return cachedRepeat

    }
}

File: hdd/longest-substring-without-repeating-characters/answer.dart
class Solution {
  int lengthOfLongestSubstring(String s) {
    int cachedRepeat = 0;

    Set<String> usingCharacters = {};
    List<String> cachedString = [];

    for (String char in s.split("")) {
      cachedString.add(char);

      if (!usingCharacters.add(char)) {
        cachedString = cachedString.sublist(cachedString.indexOf(char) + 1);
        usingCharacters = cachedString.toSet();
      }

      if (cachedRepeat < cachedString.length) {
        cachedRepeat = cachedString.length;
      }
    }
    return cachedRepeat;
  }
}
